include/boost/url/impl/url_base.ipp

```diff
@@ -1337,6 +1337,166 @@ end() const noexcept ->
     return iterator(v_, true);
 }
 
+auto
+url_base::
+segments_type::
+erase( url_base::segments_type::iterator pos ) noexcept ->
+    iterator
+{
+    auto last = pos;
+    return erase(pos, ++last);
+}
+
+auto
+url_base::
+segments_type::
+erase( url_base::segments_type::iterator first, url_base::segments_type::iterator last ) noexcept ->
+    iterator
+{
+    BOOST_ASSERT(v_ != nullptr);
+    url_base & v = *v_;
+    BOOST_ASSERT(v.size() == v.a_.size());
+    BOOST_ASSERT(first.v_ == &v);
+    BOOST_ASSERT(last.v_ == &v);
+    BOOST_ASSERT(first.off_ >= v.pt_.offset[detail::id_path]);
+    BOOST_ASSERT(last.off_ >= v.pt_.offset[detail::id_path]);
+    BOOST_ASSERT(first.off_ <= v.pt_.offset[detail::id_query]);
+    BOOST_ASSERT(last.off_ <= v.pt_.offset[detail::id_query]);
+    auto const d = last.off_ - first.off_;
+    if( d == 0 )
+        return first;
+    BOOST_ASSERT(d > 0);
+    int c = 0;
+    for( auto i = v.s_ + first.off_, e = v.s_ + last.off_; i != e; ++i )
+        c += (*i == '/');
+    BOOST_ASSERT(c > 0);
+    BOOST_ASSERT(v.pt_.nseg >= c);
+    v.pt_.nseg -= c;
+    std::memmove(v.s_ + first.off_, v.s_ + last.off_, v.pt_.offset[detail::id_end] - last.off_ + 1);
+    v.pt_.resize(detail::id_path, v.pt_.length(detail::id_path, detail::id_query) - d);
+    BOOST_ASSERT(v.size() + d == v.a_.size());
+    v.a_.resize(v.size());
+    BOOST_ASSERT(v.s_[v.pt_.offset[detail::id_end]] == '\0');
+    first.parse();
+    return first;
+}
+
+auto
+url_base::
+segments_type::
+insert_encoded_segment_( segments_type::iterator pos, string_view s ) ->
```

Comment:
instead of a trailing underscore, use the suffix `_impl` to denote that a function is a private implementation function